### ISSUE TITLE
Split CLI out into a module

### ISSUE THREAD

**Issue by contra** [gulpjs/gulp#827]
this will clean up the codebase

repo: https://github.com/gulpjs/gulp-cli

**Comment by srcspider** [comment]
@contra @phated 

I've run recently into a case where building a gulpfile was far more useful then using a makefile; for flexibility reasons, mainly because as a programmer I'd rather just take 5min to code something up then go though the trouble (ie. hours) of finding out what arcane black magic is required in the dumb makefile format. (That and it's easier to just get stuff to run in parallel using javascript) 

However that does pose the issue that setting up gulp requires a lot of fiddly work. It would be nice to have gulp work globally fine too, for the cases where you just need a gulp task system along with 1-2 very specialized modules.

My recommendation is to provide a *gulp3*, *gulp4*, etc npm module for all major breaking versions that would be specifically intended to run globally only. You would just need to forbid it from running if it's in a project with both a `node_modules` and `package.json` file alongside the `gulpfile.js` to avoid misuse. 

Having those you can have setups that are this simple and clean:

```sh
sudo npm i -g gulp3
touch gulpfile.js
```

This would be very nice for non-javascript/web projects that would literally only be using javascript/gulp to power a rudimentary build system; since there's no `package.json`, and no `node_modules`.

**Comment by contra** [comment]
@srcspider The gulp cli will work with every version of gulp, it won't matter what version you have installed.

**Comment by srcspider** [comment]
I'm referring to the case where you want to have gulp globally ONLY,  so no
package.json, node_modules etc.

Even if code like this, would work:

npm i -g gulp
npm i -g gulp-cli

It's undesirable since you end up with "this script requires gulp 3.x, but
this other project requires gulp 4.x" so it's all a mess. At the global
level unless it can work for all instances though systems like gulp-cli
does (which in this case it cant since anything more then a single gulpfile
is undesirable), its much better to be uniquely defined so it never breaks.

**Comment by heikki** [comment]
@srcspider How would any require work in that gulpfile if there's no local node_modules?

**Comment by contra** [comment]
@srcspider So you're saying the global CLI, when a local isn't found, should inject gulp somehow into the `require` system in node to fake that module being installed?

**Comment by phated** [comment]
I think @contra is saying that the request is way out of scope for gulp as it would be messing with node core functionality. We have no plans to support this type of request.

**Comment by stringparser** [comment]
@srcspider you can use [`NODE_ENV` variable to include it](http://nodejs.org/api/modules.html#modules_loading_from_the_global_folders) to include the global `node_modules` on the `require` lookup

```
NODE_PATH=$(npm config get prefix)/lib/node_modules
```

**Comment by heikki** [comment]
Yes, for example [nvm](https://github.com/Alex7Kom/nvm-fish) seems to set NODE_PATH out of the box. If there's no local module require tries global. I was **a bit** surprised when I didn't see the error I was expecting.

(Just for fun) I got plain gulpfile working by commenting some checks out from the global gulp and requiring just 'gulp' if `env.modulePath` was missing. I tried with gulp-cli too but gave up before anything worked..

**Comment by stringparser** [comment]
The missing local version error right? I tend to use gulp globally, either with the above or doing `npm link gulp`

**Comment by heikki** [comment]
Yes that. I don't know how sane that is but.. I found [one line change](https://github.com/tkellen/node-liftoff/blob/23947735ccc52a106a391cf6d644caf8862b3a14/index.js#L89) in liftoff that makes global thing work:

    modulePath = resolve.sync(this.moduleName, {basedir: configBase || cwd, paths: process.env.NODE_PATH});

Resolve [docs](https://github.com/substack/node-resolve#resolvesyncpkg-opts) say:
> opts.paths - require.paths array to use if nothing is found on the normal node_modules recursive walk (probably don't use this)

**Comment by stringparser** [comment]
imho that would make more sense for the cli instead of using one of the methods I posted above, but I'm also happy with the way things are at the moment. Ideally for me, a `gulpfile.js` could check if the script was run from its bin i.e. `which gulp` its at `process.argv` and launch the cli

**Comment by phated** [comment]
This has been split out.  Should we close or wait until gulpjs/gulp-cli#7 is closed before releasing the 3.x branch with this change?

**Comment by contra** [comment]
@phated Yeah, I think that would be best. Are the changes for this already on master, but unpublished?

**Comment by phated** [comment]
@contra gulp-cli version 0.1.3 is the same code as I pulled from gulp 3.8.x and the changes have been made for master.  I have the 4.0 stuff in a 4.0 branch on gulp-cli and I'd actually like to merge it over into gulp 3.9 so people will have automatic 4.0 compat.  I think another 3.8.x release will need to happen with the current CLI after gulpjs/gulp-cli#7 is closed just to make sure we don't have anything really bad happening.

**Comment by contra** [comment]
@phated Yeah a 3.9 seems good to me. Ping me when stuff is ready to deploy :+1:

**Comment by contra** [comment]
This is done, yeah?

**Comment by phated** [comment]
Not fully :frowning:

**Comment by contra** [comment]
closing this since it has been implemented, just waiting on npm to fix some bugs before we can publish
